File: python/src/train/ops/concat_hash.py
from __future__ import annotations

"""融合预处理算子：多输入拼接后直接 hash。"""

from typing import Any

from . import register_op
from .feature_hash import FeatureHash


@register_op("ConcatHash")
class ConcatHash:
    def __init__(
        self,
        vocab_size: int,
        *,
        separator: str = "_",
        num_hashes: int = 1,
        namespace: str = "",
        salt: str = "",
        version: str = "",
    ) -> None:
        self._hash = FeatureHash(
            vocab_size,
            num_hashes=num_hashes,
            separator=separator,
            namespace=namespace,
            salt=salt,
            version=version,
        )

    @classmethod
    def from_config(cls, params: dict) -> "ConcatHash":
        return cls(
            vocab_size=int(params.get("vocab_size", 1000)),
            num_hashes=int(params.get("num_hashes", 1)),
            separator=str(params.get("separator", "_")),
            namespace=str(params.get("namespace", "")),
            salt=str(params.get("salt", "")),
            version=str(params.get("version", "")),
        )

    def process(self, inputs: list[Any]) -> int | list[int]:
        return self._hash.process(inputs)

    def process_batch(self, inputs: list[Any]) -> list[int] | list[list[int]]:
        return self._hash.process_batch(inputs)

File: src/feats/ops/concat_hash.rs
//! 融合预处理算子：多输入拼接后直接 hash。

use crate::feats::ops::{feature_hash::FeatureHash, CustomOp, Fv};

/// 多输入字符串拼接后合并哈希。
pub struct ConcatHash {
    separator: String,
    inner: FeatureHash,
}

impl ConcatHash {
    #[allow(clippy::too_many_arguments)]
    /// 创建 ConcatHash 算子。
    pub fn new(
        vocab_size: u32,
        num_hashes: u32,
        separator: String,
        namespace: String,
        salt: String,
        version: String,
    ) -> Result<Self, String> {
        Ok(Self {
            separator: separator.clone(),
            inner: FeatureHash::with_scope(
                vocab_size, num_hashes, separator, &namespace, &salt, &version,
            )?,
        })
    }
}

pub fn create(params: &serde_yaml::Value) -> Result<Box<dyn CustomOp>, String> {
    let vocab_size = params.get("vocab_size").and_then(|v| v.as_u64()).unwrap_or(1000) as u32;
    let num_hashes = params.get("num_hashes").and_then(|v| v.as_u64()).unwrap_or(1) as u32;
    let separator = params.get("separator").and_then(|v| v.as_str()).unwrap_or("_").to_string();
    let namespace = params.get("namespace").and_then(|v| v.as_str()).unwrap_or("").to_string();
    let salt = params.get("salt").and_then(|v| v.as_str()).unwrap_or("").to_string();
    let version = params.get("version").and_then(|v| v.as_str()).unwrap_or("").to_string();
    Ok(Box::new(ConcatHash::new(vocab_size, num_hashes, separator, namespace, salt, version)?))
}

impl CustomOp for ConcatHash {
    fn name(&self) -> &str {
        "ConcatHash"
    }

    fn process(&self, inputs: &[Fv]) -> Result<Fv, String> {
        let key: String = inputs
            .iter()
            .map(|v| v.to_string())
            .collect::<Vec<_>>()
            .join(&self.separator);
        self.inner.process(&[Fv::Str(key)])
    }

    fn process_batch(&self, inputs: &[&[Fv]], n_rows: usize) -> Result<Vec<Fv>, String> {
        if n_rows == 0 {
            return Ok(vec![]);
        }
        let mut results: Vec<Fv> = Vec::with_capacity(n_rows);
        for row in 0..n_rows {
            let key: String = inputs
                .iter()
                .map(|col| {
                    if row < col.len() {
                        col[row].to_string()
                    } else {
                        String::new()
                    }
                })
                .collect::<Vec<_>>()
                .join(&self.separator);
            results.push(self.inner.process(&[Fv::Str(key)])?);
        }
        Ok(results)
    }
}
